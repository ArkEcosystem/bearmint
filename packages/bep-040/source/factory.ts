/**
 * @see https://github.com/bitcoin/bips/blob/master/bip-0013.mediawiki
 */

import { ripemd160, sha256 } from '@bearmint/bep-009'
import type {
	Address,
	AddressFactory,
	KeyPairFactory,
	PrivateKey,
	PublicKey,
} from '@bearmint/bep-013'
import { base58check } from '@scure/base'

export function makeAddress(prefix: number, bytes: Uint8Array): Address {
	return {
		toBytes() {
			return bytes
		},
		toString() {
			return base58check((data: Uint8Array) => sha256(data)).encode(bytes)
		},
		async verify() {
			return bytes[0] === prefix
		},
	}
}

export function makeAddressFactory(keyPairFactory: KeyPairFactory, prefix: number): AddressFactory {
	async function fromBytes(bytes: Uint8Array) {
		return makeAddress(prefix, bytes)
	}

	async function fromPublicKey(publicKey: PublicKey) {
		const buffer: Buffer = Buffer.from(ripemd160(publicKey.toBytes()))
		const payload: Buffer = Buffer.alloc(21)

		payload.writeUInt8(prefix, 0)
		buffer.copy(payload, 1)

		return fromBytes(payload)
	}

	return {
		fromBytes,
		async fromMnemonic(mnemonic: string) {
			return fromPublicKey((await keyPairFactory.fromMnemonic(mnemonic)).toPublicKey())
		},
		async fromPrivateKey(privateKey: PrivateKey) {
			return fromPublicKey(privateKey.toPublicKey())
		},
		fromPublicKey,
		async fromString(address: string) {
			return fromBytes(base58check((data: Uint8Array) => sha256(data)).decode(address))
		},
	}
}
