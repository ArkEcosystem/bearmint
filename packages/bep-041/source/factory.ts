/**
 * @see https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
 */

import type {
	Address,
	AddressFactory,
	KeyPairFactory,
	PrivateKey,
	PublicKey,
} from '@bearmint/bep-013'
import { bech32 } from '@scure/base'

export function makeAddress(prefix: string, bytes: Uint8Array): Address {
	function toString() {
		return bech32.encode(prefix, bech32.toWords(bytes))
	}

	return {
		toBytes() {
			return bytes
		},
		toString,
		async verify() {
			try {
				bech32.decode(toString())

				return true
			} catch {
				return false
			}
		},
	}
}

export function makeAddressFactory(keyPairFactory: KeyPairFactory, prefix: string): AddressFactory {
	async function fromBytes(bytes: Uint8Array) {
		return makeAddress(prefix, bytes)
	}

	async function fromPublicKey(publicKey: PublicKey) {
		return makeAddress(prefix, publicKey.toBytes())
	}

	return {
		fromBytes,
		async fromMnemonic(mnemonic: string) {
			return fromPublicKey((await keyPairFactory.fromMnemonic(mnemonic)).toPublicKey())
		},
		async fromPrivateKey(privateKey: PrivateKey) {
			return fromPublicKey(privateKey.toPublicKey())
		},
		fromPublicKey,
		async fromString(address: string) {
			return fromBytes(bech32.decodeToBytes(address).bytes)
		},
	}
}
