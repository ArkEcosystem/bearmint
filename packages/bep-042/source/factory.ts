/**
 * @see https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki
 */

import type {
	Address,
	AddressFactory,
	KeyPairFactory,
	PrivateKey,
	PublicKey,
} from '@bearmint/bep-013'
import type { Bech32DecodedWithArray } from '@scure/base'
import { bech32m } from '@scure/base'

export function makeAddress(prefix: string, decoded: Bech32DecodedWithArray): Address {
	function toString() {
		return bech32m.encode(decoded.prefix, bech32m.toWords(decoded.bytes))
	}

	return {
		toBytes() {
			return decoded.bytes
		},
		toString,
		async verify() {
			return decoded.prefix === prefix
		},
	}
}

export function makeAddressFactory(keyPairFactory: KeyPairFactory, prefix: string): AddressFactory {
	async function fromBytes(bytes: Uint8Array) {
		return makeAddress(prefix, {
			bytes,
			prefix,
			words: bech32m.toWords(bytes),
		})
	}

	async function fromPublicKey(publicKey: PublicKey) {
		return fromBytes(publicKey.toBytes())
	}

	return {
		fromBytes,
		async fromMnemonic(mnemonic: string) {
			return fromPublicKey((await keyPairFactory.fromMnemonic(mnemonic)).toPublicKey())
		},
		async fromPrivateKey(privateKey: PrivateKey) {
			return fromPublicKey(privateKey.toPublicKey())
		},
		fromPublicKey,
		async fromString(address: string) {
			return makeAddress(prefix, bech32m.decodeToBytes(address))
		},
	}
}
