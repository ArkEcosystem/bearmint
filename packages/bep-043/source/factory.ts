import { bytesToHex, hexToBytes } from '@bearmint/bep-009'
import type {
	Address,
	AddressFactory,
	KeyPairFactory,
	PrivateKey,
	PublicKey,
} from '@bearmint/bep-013'
import { isValidAddress, privateToAddress, publicToAddress } from '@ethereumjs/util'

export function makeAddress(bytes: Uint8Array): Address {
	function toString() {
		return `0x${bytesToHex(bytes)}`
	}

	return {
		toBytes() {
			return bytes
		},
		toString,
		async verify() {
			return isValidAddress(toString())
		},
	}
}

export function makeAddressFactory(keyPairFactory: KeyPairFactory): AddressFactory {
	return {
		async fromBytes(bytes: Uint8Array) {
			return makeAddress(bytes)
		},
		async fromMnemonic(mnemonic: string) {
			return makeAddress(
				publicToAddress(
					Buffer.from((await keyPairFactory.fromMnemonic(mnemonic)).toPublicKey().toBytes()),
					true,
				),
			)
		},
		async fromPrivateKey(privateKey: PrivateKey) {
			return makeAddress(privateToAddress(Buffer.from(privateKey.toBytes())))
		},
		async fromPublicKey(publicKey: PublicKey) {
			return makeAddress(publicToAddress(Buffer.from(publicKey.toBytes()), true))
		},
		async fromString(address: string) {
			if (address.toLowerCase().startsWith('0x')) {
				address = address.slice(2)
			}

			return makeAddress(hexToBytes(address))
		},
	}
}
