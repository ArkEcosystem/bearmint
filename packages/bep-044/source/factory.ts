/* eslint-disable unicorn/consistent-function-scoping */
/**
 * @see https://github.com/tendermint/tendermint/blob/master/crypto/secp256k1/secp256k1.go#L148-L161
 */

import { bytesToHex, hexToBytes, ripemd160, sha256 } from '@bearmint/bep-009'
import type {
	Address,
	AddressFactory,
	KeyPairFactory,
	PrivateKey,
	PublicKey,
} from '@bearmint/bep-013'

export function makeAddress(bytes: Uint8Array): Address {
	return {
		toBytes() {
			return bytes
		},
		toString() {
			return bytesToHex(bytes)
		},
		async verify() {
			return bytes.length === 20
		},
	}
}

export function makeAddressFactory(keyPairFactory: KeyPairFactory): AddressFactory {
	async function fromBytes(bytes: Uint8Array) {
		return makeAddress(bytes)
	}

	async function fromPublicKey(publicKey: PublicKey) {
		return makeAddress(ripemd160(sha256(publicKey.toBytes())))
	}

	return {
		fromBytes,
		async fromMnemonic(mnemonic: string) {
			return fromPublicKey((await keyPairFactory.fromMnemonic(mnemonic)).toPublicKey())
		},
		async fromPrivateKey(privateKey: PrivateKey) {
			return fromPublicKey(privateKey.toPublicKey())
		},
		fromPublicKey,
		async fromString(address: string) {
			return fromBytes(hexToBytes(address))
		},
	}
}
