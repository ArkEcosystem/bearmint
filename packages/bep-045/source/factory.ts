/**
 * @see https://polkadot.js.org/docs/keyring/start/ss58/
 * @see https://docs.substrate.io/v3/advanced/ss58/
 */

import type {
	Address,
	AddressFactory,
	KeyPairFactory,
	PrivateKey,
	PublicKey,
} from '@bearmint/bep-013'
import { decodeAddress, encodeAddress } from '@polkadot/util-crypto'

export function makeAddress(prefix: number, bytes: Uint8Array): Address {
	return {
		toBytes() {
			return bytes
		},
		toString() {
			return encodeAddress(bytes, prefix)
		},
		async verify() {
			try {
				decodeAddress(encodeAddress(bytes, prefix), undefined, prefix)

				return true
			} catch {
				return false
			}
		},
	}
}

export function makeAddressFactory(keyPairFactory: KeyPairFactory, prefix: number): AddressFactory {
	async function fromBytes(bytes: Uint8Array) {
		return makeAddress(prefix, bytes)
	}

	async function fromPublicKey(publicKey: PublicKey) {
		return fromBytes(publicKey.toBytes())
	}

	return {
		fromBytes,
		async fromMnemonic(mnemonic: string) {
			return fromPublicKey((await keyPairFactory.fromMnemonic(mnemonic)).toPublicKey())
		},
		async fromPrivateKey(privateKey: PrivateKey) {
			return fromPublicKey(privateKey.toPublicKey())
		},
		fromPublicKey,
		async fromString(address: string) {
			return fromBytes(decodeAddress(address))
		},
	}
}
