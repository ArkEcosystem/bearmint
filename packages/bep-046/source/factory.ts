/* eslint-disable unicorn/consistent-function-scoping */
/**
 * @see https://github.com/tendermint/tendermint/blob/master/crypto/ed25519/ed25519.go#L160-L166
 */

import { bytesToHex, hexToBytes, sha256 } from '@bearmint/bep-009'
import type {
	Address,
	AddressFactory,
	KeyPairFactory,
	PrivateKey,
	PublicKey,
} from '@bearmint/bep-013'

export function makeAddress(bytes: Uint8Array): Address {
	return {
		toBytes() {
			return bytes
		},
		toString() {
			return bytesToHex(bytes)
		},
		async verify() {
			return bytes.byteLength === 20
		},
	}
}

export function makeAddressFactory(keyPairFactory: KeyPairFactory): AddressFactory {
	async function fromBytes(bytes: Uint8Array) {
		return makeAddress(bytes)
	}

	async function fromPublicKey(publicKey: PublicKey) {
		return makeAddress(sha256(publicKey.toBytes()).slice(0, 20))
	}

	return {
		fromBytes,
		async fromMnemonic(mnemonic: string) {
			return fromPublicKey((await keyPairFactory.fromMnemonic(mnemonic)).toPublicKey())
		},
		async fromPrivateKey(privateKey: PrivateKey) {
			return fromPublicKey(privateKey.toPublicKey())
		},
		fromPublicKey,
		async fromString(address: string) {
			return fromBytes(hexToBytes(address))
		},
	}
}
