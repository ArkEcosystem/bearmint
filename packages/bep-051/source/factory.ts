import { assert, blake3, bytesToHex, sha256 } from '@bearmint/bep-009'
import type { Cradle, TxFactory } from '@bearmint/bep-013'
import { PROTO_OPTS } from '@bearmint/bep-013'
import type { TxMsg } from '@bearmint/bep-018'
import { SignedTx, TxChecksum } from '@bearmint/bep-018'

export function makeTxFactory(
	cradle: Pick<Cradle, 'AddressFactory' | 'PublicKeyFactory'>,
): TxFactory {
	return {
		async fromBytes(bytes: Uint8Array) {
			const tx = SignedTx.fromBinary(bytes, PROTO_OPTS)

			assert.defined<TxMsg>(tx.message)

			return Object.seal({
				bytes,
				checksum: TxChecksum.fromJson({
					handler: tx.message.handler,
					network: tx.message.network,
					version: tx.message.version,
				}),
				data: {
					gas: BigInt(tx.gas),
					message: tx.message,
					nonce: BigInt(tx.nonce),
					recipient: await cradle.AddressFactory.fromString(tx.recipient),
					sender: await cradle.PublicKeyFactory.fromString(tx.sender),
					signature: tx.signature,
					version: tx.version,
				},
				/**
				 * @see https://github.com/tendermint/tendermint/blob/main/types/tx.go#L28-L31
				 * @see https://github.com/tendermint/tendermint/blob/main/crypto/tmhash/hash.go#L18-L22
				 */
				hash: bytesToHex(sha256(bytes)),
				raw: tx,
				type: bytesToHex(blake3(new TxChecksum(tx.message).toBinary())),
			})
		},
	}
}
