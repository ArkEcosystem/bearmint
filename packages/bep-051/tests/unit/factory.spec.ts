import { expect, describe, it } from 'vitest'
import { extendExpect, spyFn } from '@bearmint/bep-005'
extendExpect(expect)
import { hexToBytes } from '@bearmint/bep-009'
import { SignedTx } from '@bearmint/bep-018'

import { makeTxFactory } from '../../source/factory.js'

describe('Tx Factory', () => {
	it('should create the context functions', async () => {
		expect(
			await makeTxFactory({
				AddressFactory: { fromString: spyFn().toFunction() },
				PublicKeyFactory: { fromString: spyFn().toFunction() },
			}).fromBytes(
				new SignedTx({
					gas: BigInt(1),
					hash: '134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e',
					message: {
						content: Buffer.alloc(0),
						handler: 'handler',
						network: 'network',
						version: 'version',
					},
					nonce: BigInt(1),
					recipient: 'recipient',
					sender: '134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e',
					signature: hexToBytes(
						'134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e',
					),
					version: '1.0.0',
				}).toBinary(),
			),
		).toMatchSnapshot()
	})
})
