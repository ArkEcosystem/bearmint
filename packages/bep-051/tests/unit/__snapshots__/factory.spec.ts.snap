// Vitest Snapshot v1

exports[`Tx Factory > should create the context functions 1`] = `
{
  "bytes": Uint8Array [
    10,
    5,
    49,
    46,
    48,
    46,
    48,
    18,
    64,
    49,
    51,
    52,
    97,
    51,
    99,
    99,
    99,
    51,
    51,
    51,
    53,
    48,
    55,
    49,
    55,
    97,
    55,
    100,
    98,
    55,
    57,
    97,
    57,
    51,
    98,
    53,
    53,
    49,
    54,
    101,
    97,
    48,
    97,
    53,
    100,
    52,
    100,
    98,
    48,
    97,
    54,
    48,
    102,
    48,
    51,
    97,
    48,
    56,
    51,
    51,
    53,
    101,
    48,
    99,
    52,
    99,
    57,
    55,
    53,
    100,
    49,
    53,
    101,
    26,
    9,
    114,
    101,
    99,
    105,
    112,
    105,
    101,
    110,
    116,
    32,
    1,
    40,
    1,
    50,
    27,
    10,
    7,
    104,
    97,
    110,
    100,
    108,
    101,
    114,
    18,
    7,
    118,
    101,
    114,
    115,
    105,
    111,
    110,
    26,
    7,
    110,
    101,
    116,
    119,
    111,
    114,
    107,
    58,
    64,
    19,
    74,
    60,
    204,
    51,
    53,
    7,
    23,
    167,
    219,
    121,
    169,
    59,
    85,
    22,
    234,
    10,
    93,
    77,
    176,
    166,
    15,
    3,
    160,
    131,
    53,
    224,
    196,
    201,
    117,
    209,
    94,
    19,
    74,
    60,
    204,
    51,
    53,
    7,
    23,
    167,
    219,
    121,
    169,
    59,
    85,
    22,
    234,
    10,
    93,
    77,
    176,
    166,
    15,
    3,
    160,
    131,
    53,
    224,
    196,
    201,
    117,
    209,
    94,
  ],
  "checksum": TxChecksum {
    "handler": "handler",
    "network": "network",
    "version": "version",
  },
  "data": {
    "gas": 1n,
    "message": TxMsg {
      "content": Uint8Array [],
      "handler": "handler",
      "network": "network",
      "version": "version",
    },
    "nonce": 1n,
    "recipient": undefined,
    "sender": undefined,
    "signature": Uint8Array [
      19,
      74,
      60,
      204,
      51,
      53,
      7,
      23,
      167,
      219,
      121,
      169,
      59,
      85,
      22,
      234,
      10,
      93,
      77,
      176,
      166,
      15,
      3,
      160,
      131,
      53,
      224,
      196,
      201,
      117,
      209,
      94,
      19,
      74,
      60,
      204,
      51,
      53,
      7,
      23,
      167,
      219,
      121,
      169,
      59,
      85,
      22,
      234,
      10,
      93,
      77,
      176,
      166,
      15,
      3,
      160,
      131,
      53,
      224,
      196,
      201,
      117,
      209,
      94,
    ],
    "version": "1.0.0",
  },
  "hash": "c373b61536afcad057e0adadae367a2f62d778e0ee9e00549f4772fa4849c499",
  "raw": SignedTx {
    "gas": 1n,
    "message": TxMsg {
      "content": Uint8Array [],
      "handler": "handler",
      "network": "network",
      "version": "version",
    },
    "nonce": 1n,
    "recipient": "recipient",
    "sender": "134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e",
    "signature": Uint8Array [
      19,
      74,
      60,
      204,
      51,
      53,
      7,
      23,
      167,
      219,
      121,
      169,
      59,
      85,
      22,
      234,
      10,
      93,
      77,
      176,
      166,
      15,
      3,
      160,
      131,
      53,
      224,
      196,
      201,
      117,
      209,
      94,
      19,
      74,
      60,
      204,
      51,
      53,
      7,
      23,
      167,
      219,
      121,
      169,
      59,
      85,
      22,
      234,
      10,
      93,
      77,
      176,
      166,
      15,
      3,
      160,
      131,
      53,
      224,
      196,
      201,
      117,
      209,
      94,
    ],
    "version": "1.0.0",
  },
  "type": "900578960398f70afd2c6c71d499250e46c147dcccc14a1dd5681347852cdefe",
}
`;
